statefun-testutil/src/main/java/org/apache/flink/statefun/testutils/function/TestContext.java

```diff
@@ -0,0 +1,198 @@
+
+  private long watermark;
+
+  TestContext(Address selfAddress, StatefulFunction function, Instant startTime) {
+    this.selfAddress = Objects.requireNonNull(selfAddress);
+    this.function = function;
+
+    this.watermark = startTime.toEpochMilli();
+    this.messages = new ArrayDeque<>();
+    this.pendingMessage = new PriorityQueue<>(Comparator.comparingLong(a -> a.timer));
+    this.outputs = new HashMap<>();
+  }
+
+  @Override
+  public Address self() {
+    return selfAddress;
+  }
+
+  @Override
+  public Address caller() {
+    return from;
+  }
+
+  @Override
+  public void reply(Object message) {
+    Address to = caller();
+    if (to == null) {
+      throw new IllegalStateException("The caller address is null");
+    }
+
+    send(to, message);
+  }
+
+  @Override
+  public void send(Address to, Object message) {
+    if (to.equals(selfAddress)) {
+      messages.add(new Envelope(self(), to, message));
+      return;
+    }
+    responses.computeIfAbsent(to, ignore -> new ArrayList<>()).add(message);
+  }
+
+  @Override
+  public <T> void send(EgressIdentifier<T> egress, T message) {
+    outputs.computeIfAbsent(egress, ignore -> new ArrayList<>()).add(message);
+  }
+
+  @Override
+  public void sendAfter(Duration delay, Address to, Object message) {
+    pendingMessage.add(
+        new PendingMessage(new Envelope(self(), to, message), watermark + delay.toMillis()));
+  }
+
+  @Override
+  public <M, T> void registerAsyncOperation(M metadata, CompletableFuture<T> future) {
+    T value = null;
+    Throwable error = null;
+
+    try {
+      value = future.get();
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Failed to get results from async action", e);
+    } catch (ExecutionException e) {
+      error = e.getCause();
+    }
+
+    AsyncOperationResult.Status status;
+    if (error == null) {
+      status = AsyncOperationResult.Status.SUCCESS;
+    } else {
+      status = AsyncOperationResult.Status.FAILURE;
+    }
+
+    AsyncOperationResult<M, T> result = new AsyncOperationResult<>(metadata, status, value, error);
+    messages.add(new Envelope(self(), self(), result));
+  }
+
+  @SuppressWarnings("unchecked")
+  <T> List<T> getEgress(EgressIdentifier<T> identifier) {
+    List<?> values = outputs.getOrDefault(identifier, Collections.emptyList());
+
+    // Because the type is part of the identifier key
+    // this cast is always safe.
+    return (List<T>) values;
+  }
+
+  Map<Address, List<Object>> invoke(Address from, Object message) {
+    messages.add(new Envelope(from, null, message));
+
+    return processAllMessages();
+  }
+
+  Map<Address, List<Object>> tick(Duration duration) {
+    watermark += duration.toMillis();
+
+    while (!pendingMessage.isEmpty() && pendingMessage.peek().timer <= watermark) {
+      messages.add(pendingMessage.poll().envelope);
+    }
+
+    return processAllMessages();
+  }
+
+  private Map<Address, List<Object>> processAllMessages() {
```

Comment:
mostly a nitpick:

I feel like the responsibility of this method should be separated from the `TestContext`, perhaps a `SynchronousReductions` class, which uses the `TestContext`.
Adding to message queue and draining it will reside in the `SynchronousReductions`. 

I think it's good as it it is though, so will still merge this as it. Just food for thought.